src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeWriter.cs

```diff
@@ -0,0 +1,316 @@
+
+        private BufferSegment _head;
+        private BufferSegment _tail;
+        private Memory<byte> _tailMemory;
+        private int _tailBytesBuffered;
+        private int _bytesBuffered;
+
+        private MemoryPool<byte> _pool;
+
+        private CancellationTokenSource _internalTokenSource;
+        private bool _isCompleted;
+        private object _lockObject = new object();
+
+        private BufferSegmentStack _bufferSegmentPool;
+
+        private CancellationTokenSource InternalTokenSource
+        {
+            get
+            {
+                lock (_lockObject)
+                {
+                    if (_internalTokenSource == null)
+                    {
+                        _internalTokenSource = new CancellationTokenSource();
+                    }
+                    return _internalTokenSource;
+                }
+            }
+        }
+
+        public StreamPipeWriter(Stream writingStream, StreamPipeWriterOptions options)
+        {
+            InnerStream = writingStream ?? throw new ArgumentNullException(nameof(writingStream));
+
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            _minimumBufferSize = options.MinimumBufferSize;
+            _pool = options.Pool == MemoryPool<byte>.Shared ? null : options.Pool;
+            _bufferSegmentPool = new BufferSegmentStack(InitialSegmentPoolSize);
+        }
+
+        /// <summary>
+        /// Gets the inner stream that is being written to.
+        /// </summary>
+        public Stream InnerStream { get; }
+
+        /// <inheritdoc />
+        public override void Advance(int bytes)
+        {
+            if ((uint)bytes > (uint)_tailMemory.Length)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.bytes);
+            }
+
+            _tailBytesBuffered += bytes;
+            _bytesBuffered += bytes;
+            _tailMemory = _tailMemory.Slice(bytes);
+        }
+
+        /// <inheritdoc />
+        public override Memory<byte> GetMemory(int sizeHint = 0)
+        {
+            if (_isCompleted)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoWritingAllowed();
+            }
+
+            if (sizeHint < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.sizeHint);
+            }
+
+            AllocateMemory(sizeHint);
+
+            return _tailMemory;
+        }
+
+        /// <inheritdoc />
+        public override Span<byte> GetSpan(int sizeHint = 0)
+        {
+            if (_isCompleted)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoWritingAllowed();
+            }
+
+            if (sizeHint < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.sizeHint);
+            }
+
+            AllocateMemory(sizeHint);
```

Comment:
Consider adding a fast path here when `_tailMemory.Length >= sizeHint`

Reply:
I didn't plan to do any performance work until the next preview. I need to port the performance tests etc. I just want to get in the new API with implementation and tests for preview4.